### DISCUSSION TITLE
Extendable UI Concept

### DISCUSSION THREAD

**Discussion by devmount** [devmount/third-stats#376] [Ideas]
ThirdStats started as [a simple one-pager containing 6 charts](https://user-images.githubusercontent.com/5441654/93931445-14066b80-fd1f-11ea-8fe7-bde8674b26f4.png). Since then, ThirdStats evolved to a much more complex analysis tool taking many different metrics into account and providing account comparison, filtering and more. To prevent the current ThirdStats UI of the stats page to become more and more confusing, there needs to be a UI concept that allows the addition of more chart, chart sections and filter fields while maintaining clarity and intuitive handling.

In the following are some proposals to address the above mentioned issues.

## 1. Stats Categories

The stats page can be divided into different groups. This way, the page won't get overloaded and the different charts are better organized. Here is a draft of a possible tab menu:

![image](https://user-images.githubusercontent.com/5441654/170839549-5311c2d9-621b-41f0-8248-3219c3d65f41.png)

The tabs indicate the currently active subpage, adapting the concept and style of the Thunderbird tabs and colors. Also this tab menu should be sticky when scrolling down.

## 2. Filter menu

There is currently the account selection with refresh button, 3 filter fields (folder, date range, contact) and 2 icon buttons (export, options). Filter fields and icon buttons can be moved into a drawer, which slides in e.g. from the right when clicking a menu icon in the upper right. Here is a draft of a possible drawer:

![thirdstats_concept_drawer](https://user-images.githubusercontent.com/5441654/170945757-e1571e68-61e2-43d1-b07d-83699bfc610a.png)

There should be an option to pin the drawer for filter power users. Also there should be an indication if one or more filters are active and the drawer is closed.

## 3. Compress Page Header

Applying the changes described above gives the possibility to compress the page header to a smaller height, giving more space for the actual content. Here is a draft for a possible change:

![thirdstats_concept_ui_change](https://user-images.githubusercontent.com/5441654/170844414-757d1841-088d-4c76-a49b-357eeb062730.png)

## Your Feedback

These changes would allow to be much more flexible in adding charts, filter fields or other elements.

I would love to hear your thoughts about ThirdStats UI in general as well as about the above mentioned proposals. Let's discuss, how we can make ThirdStats look and feel even better together. Feel free to use the following questions as starting point:

- How satisfied are you with the current UI? What bothers you the most?
- Would a tab navigation of different stats categories be an improvement or more of a hindrance for you when browsing the stats?
- How often do you use the filter (folder, date range, contact)? Would you prefer having it always visible or only showing it when needed?

## Context

We already have some great discussions about UI elements, which I want to mention in this context as well to keep them in mind:

- #188
- #273 
- #279 
- #406

**Comment by koobs** [comment]
@devmount Any specific questions or considerations you either have in mind or you think are worth asking in this Discussion would be great itemized here, as primers in terms of how you'd like us to go about helping you narrow down the plan, build confidence and decide

**Reply by devmount** [reply]
Thank you for this suggestion @koobs! I added some questions to the *Your Feedback* section as starting point. Basically I need a plan to be able to add as many charts / filters as I want without making the UI confusing or revamping everything. Any constructive feedback on this helps.

**Comment by koobs** [comment]
## Summary

I think this feature/discussion is a good opportunity (and time) to review the fundamentals, really think, and distill some concepts, to lay "intentional" and "designed" ground work for the future of third-stats.

It's challenging at the best of times to 'add features and possible complexity' without making things harder for users, or losing value for some significant subset of users. And I think this is a big one of those.

**Note**: If on the other hand, you'd like to, for whatever or no reason, approach this in the reverse, as in: whats the best or simplest way to add 'extendability' to third-stats as soon as possible given only what currently exists, then let us all know so we can orient ourselves and our responses in that way.

With that in mind, and per some of the questions, thoughts and feelings below (in answer to your questions), and given we'd like this (I assume) to be a collaborative process, I think the most important thing is that we frame or sketch a 'process' for this.  This is important to:

 * Avoid this ending up a 'features' or 'details and minutiae' discussion. Not that details don't matter, but without some framing, it will be easy to get off target and lose focus. It's really tough to 'assess' anything without some lens/measure through which to do so.
 * Ensure we have good 'reasons' and motivations for 'deciding', and in particular, being able to show how decisions flow from goals/criteria (desirable properties/outcomes).

I don't necessarily *know* or want to dictate what that process might look like, either in terms of how you like or want this to work or what will work for others, but I'll try to elucidate some of my thoughts here, on that 'process':

* What do we want from a 'future third-stats'? Fundamentally. In broad strokes. These don't need to just be 'features'.
* What are some ideas/features/things we want third-stats to do/support/have going forward? I think a list, or "roadmap" would be supremely beneficial here. With that roadmap/list, what are the common elements of any of them that **inform** or **constrain** an extensibility and customization implementation?
* What are the primary data types and dimensions were working with now or would need to support? What are best practices / methods / affordances for these data types and dimensions?

## How satisfied are you with the current UI? What bothers you the most?

 * Look and feel: 8/10
 * Value/Usability: 5/10

### Lack of 'common' or 'standard' data visualization mechanics, dimensions and language *shared* among visualizations. Mechanics are specific to and designed for each visualization.
   
At the moment there's three implicit primary dimensions exposed, time, counts and 'people'. We see the use of "time" labels (weeks, quarters, years, and mail: sent and from (people)

As more metrics and/or dimensions are exposed, providing common mechanics will become increasingly important as it forms the foundation of, and "grounds" the affordances / controls that will or can be exposed to users. 

 * What are these other dimensions? How can/do they relate to each other.?

There's only so many ways to slice and dice particular data sets. Knowing/understanding one, should enable familiarity using others. 

  - Adding a specific or new "visualization" ability in the core, should enable that functionality for any graph/metric that supports that ability, based on the "data/type" of visualization it is, rather than added on an ad-hoc basis. [1]

For example: We want to add a 3D visualization. This would/could only apply to data sets with 3 (relevent) dimensions. What about if there's 4? which dimensions are best or selected?

[1] https://en.wikipedia.org/wiki/Data_and_information_visualization

###  Terminology selection and consistency. 

I'm a third-stats user from the start. I still don't always feel *entirely* comfortable when I'm looking at stats.

 * Missed opportunity to use more common/familiar/canonical mail terms.

Many examples:

 * Daytime / By Hour / Hourly
 * mails / messages
 * Received from / Sender
 * Sent to  / Recipient
 *  Flagged / Tagged / Starred
 * Temporal distribution / Hourly
 * ....

Whats the minimal set of terms possible? 

A common and consistent language/dictionary of terms will improve usability and understanding, and complement visualization mechanics and dimension-specific functionality mentioned above.
	
### "Relatively" low value insights / utility overall.
 
Eg: No ability to 'drill down' into 'first-level' stats, or 'do' anything else other than view top-level stats.

See Also: https://github.com/devmount/third-stats/issues/279

## Would a tab navigation of different stats categories be an improvement or more of a hindrance for you when browsing the stats?

This assumes categories and a specific implementation (tabs) without specifying a problem. *waves orange/red flag* :)

 * Whats the problem? Guess: 'limited space'. What else?

### Thoughts

 * How valuable/duplicative are existing graphs? Are they materially valuable, or are some just filling up space?
 * Existing graphs are "devmounts" graphs, not mine. *hint*. See Also: https://github.com/devmount/third-stats/issues/279
 * What are the categories? We should list them.
 * How do these 'categories' relate to, or are they the same as the primary dimensions of this kind of (mail) data? How many are there now? How many could there be in the future?
 * For the problem definition for which tabs may be a solution, what are alternatives?
 * Are categories (in the way you're thinking of them) the way users think about, or discover things about their mail usage?

This leads to: 

 * What is the purpose (reason for living) of third-stats? "Tagline it".
 * What problems does it solve?
 * What *unique* value/ability is it supposed to provide users?

Thoughts/feelings: 

 * Insight, discovery, understanding. What else? Is this enough?
 * How does/will third-stats empower/enable users to be easily achieve these things?

## How often do you use the filter (folder, date range, contact)? Would you prefer having it always visible or only showing it when needed?

Hardly ever. The benefit is low and use is finicky. 

* No 'date selection control'. Manual string format entry 
* Cant just type '2020' or 'now' or leave From or To empty (to imply 'all the way to the end').
* How do we know when it's needed? How will users tell you its needed if its not visible? How will users know its possible?

**Reply by devmount** [reply]
@koobs Thank you so much for taking the time for such a detailed and helpful answer! And also for supporting this project right from the start! Give me some time to process your reply, for some of your points I would like to create separate issues / discussion threads.

**Reply by koobs** [reply]
> @koobs Thank you so much for taking the time for such a detailed and helpful answer! And also for supporting this project right from the start! Give me some time to process your reply, for some of your points I would like to create separate issues / discussion threads.

My pleasure take your time :)

**Reply by devmount** [reply]
Thanks again @koobs for your thoughts. I've written on this reply over the last months when I had some time in between, so sorry in advance if some things don't stick together 😇 

## Process and Frame

Yes, I would love this to be a collaborative process which is why I started this thread in the first place. On the other hand I'm literally the only maintainer / code contributor / designer in this project atm and my time is limited. So implementing greater features or revamping the UI either takes a huge amount of time (months, years?) or we just take what currently exists and develop an extendability concept that wouldn't take too much effort (e.g. my proposals above).

### What do we want from a future ThirdStats?

For me: ThirdStats is a dashboard that gives insights about the messages I'm holding in my mail accounts and therefore insights about my behavior with messages. I'm interested in how many messages I send/receive, when I send/receive messages, from whom I receive or send messages to, how my messages are organized (folders, tags, starred), how big messages are (memory, char/word/sentence count?) and how many messages aren't useful (junk, trash). I'd like to discover information I wasn't aware of and change my behaviour about it e.g. to save time. A future ThirdStats should visualize those information organized in sections with corresponding numbers and charts. I should be able to customize this dashboard (as you already suggested in #279) to hide sections I don't need and highlight/feature section I need most.

### Roadmap: What are ideas/features/things we want ThirdStats to do/support/have going forward?

- More stats:
	- Stats about memory usage (see #213) of messages (e.g. per year/quarter/month/week/contact/folder, number of messages with attachments)
	- Stats about calendar events (this would be a whole new category which is not about messages, number of events per year/quarter/month/week per calendar, event duration)
	- Stats about tasks (also a whole new category, number of tasks per year/quarter/month/week open or completed)
- More customizability (#279)
- More immersive integration into TB (e.g. showing charts on account pages, calendar tab, tasks tab, etc.), as soon as TB allows this. Makes sense to implement this after TB 115 release

### What are the primary data types and dimensions?

Dimensions:

- Time
- Counts
- Memory size

Entities:

- Message (belonging to Folder belonging to Account)
- Contact
- Event
- Task

Entities can be given any dimension. Did I forget anything?

### Missed opportunity to use more common/familiar/canonical mail terms

> Whats the minimal set of terms possible?

I always tried to adapt to the terms Thunderbird itself is using, but it's very likely I failed that a lot, also ince I'm no native speaker. In other words: I need support on this and created #390 for anyone to contribute.

### Whats the problem, tab navigation tries to solve?

The problem is keeping ThirdStats simple an clear while adding charts. Users should be able to know at first glance, what charts/numbers they're looking at. Having all charts of all dimensions for all entities on one page (as it currently is) _can_ be a mess.

> How valuable/duplicative are existing graphs? Are they materially valuable, or are some just filling up space?

This is almost always opinionated. So if users can define the order and visibility of all charts (#279), this should be fine.

> What are the categories?

That is the interesting question indeed. The most obvious categorization would be by dimension or by entity. Both works, but categorizing by entity would make more sense, since it's already established in Thunderbird itself (see spaces toolbar).

> For the problem definition for which tabs may be a solution, what are alternatives?

The problem is structuring the content in a way that is easy to understand and navigate for the user. The only alternative I can think of is keeping it single page and define sections e.g. with anchors to jump to.

> What is the purpose (reason for living) of ThirdStats?

Provide beautifully visualized statistics about emails in Thunderbird.

> What problems does it solve?

Lack of insights about the users email behaviour.

> What unique value/ability is it supposed to provide users?

- There is no other modern statistics add-on in store.
- Beautiful and easy to understand charts.
- Privacy-respecting and Open Source. The real 'free'.

### Filter

Thank you for your honest response to this. I think it's good to have the filter for people that want to specify the stats, but I know it needs some reworking to be a lot more usable.

**Comment by Jak2k** [comment]
Maybe display a few charts on the main-page and a link in the corner of every chart to see other charts related to this chart.

**Reply by devmount** [reply]
Thank you for this suggestions. Which charts would you show on the main page? Where should the links lead to? Other dedicated pages? Which pages would that be? I like the idea of having a kind of dashboard for all kinds of charts and dedicated pages / tabs for chart categories.

**Comment by MagicFab** [comment]
Storage size missing is quite a surprise when getting stats. Thanks for everything else, just missing total size of mailbox, at the minimum.

**Reply by devmount** [reply]
Agreed. In general it's a matter of existence of a corresponding web extension API. At the time of the creation of this add-on, it was not possible to gather storage size data (without using experimental or legacy features). Now we have a [`size` attribute for MessageHeaders](https://webextension-api.thunderbird.net/en/latest/messages.html#messageheader), I just didn't have time to implement it yet. But this is definitely on the list. (total size, size distributed over time/folders/contacts etc.)

**Comment by bkmdev** [comment]
@devmount Any update on this by chance?  I too agree with @MagicFab 's comment.  The UI looks great as is 👍 and debating between these 3 mockups seems like splitting hairs (FWIW, my vote is for mockup 1 but maybe compressed vertically to remove some of the dead space between various filters and the new proposed tabs).  The key thing missing are statistics on sizes of (groups of) messages- this was one of my original motivations for installing this Add-On 🙂  

I'd love to see this Add-On become an analogue of tools like "[Grand Perspective](https://en.wikipedia.org/wiki/GrandPerspective)" which had some wonderful [treemap](https://en.wikipedia.org/wiki/Treemapping) visualizations to quickly identify large space consumers on the file system.  
![image](https://github.com/user-attachments/assets/7ca65136-afa7-4924-94ca-197ae64aa12f)
One of its great features is a way to continually "drill down" through folders all the way to the individual file level.  Color coding also added another dimension (I've usually seen it to represent file type but in this visualization it seems it is used to represent age- although aesthetically I personally would have used a mono color gradient with more grayed out items representing older items 😉  )
The point is this visualization & exploration method can 💯 % be applied to (folders of) email message(s) in Thunderbird.   

PS:  A more "modern" visualization is offered by MacOS App Store "[Disk Space Analyzer](https://apps.apple.com/us/app/disk-space-analyzer-inspector/id446243721?mt=12)" which uses a circular donut-like chart but I feel the original treemap visualization is easier to navigate although a donut chart intuitively can be used to also convey % usage (either visualization would be welcome additions to ThirdStats 🙂):  
![image](https://github.com/user-attachments/assets/3accdce0-f650-4bb4-b4c3-457e1d6ecb8f)

My 2¢ & thanks for creating this!

**Reply by devmount** [reply]
Awesome ideas! Thank you for sharing this. I would love to see this in ThirdStats. Unfortunately there is too less free time on my end and too less support from the cummunity to bring such a feature to ThirdStats at the moment. But we should definitely keep that in mind!
